swarm.go

```diff
@@ -160,59 +164,94 @@ func (c *Swarm) UpdateServices(ctx context.Context, imageName ...string) error {
 		return fmt.Errorf("failed to get service list: %w", err)
 	}
 
-	var serviceID string
+	log.Printf("found %d services", len(services))
+
+	wg := sync.WaitGroup{}
+	serviceLock := sync.Mutex{}
+
+	var swarmUpdaterService swarm.Service = swarm.Service{ID: "na"}
+	var serviceQueue []swarm.Service = make([]swarm.Service, 0)
 
+	// find swarm-updater and remove itself from the list to update last
 	for _, service := range services {
-		if c.validService(service) {
+		if _, ok := service.Spec.Annotations.Labels[serviceLabel]; ok {
+			swarmUpdaterService = service
+			continue
+		}
+
+		serviceQueue = append(serviceQueue, service)
+	}
 
-			// try to identify this service
-			if _, ok := service.Spec.Annotations.Labels[serviceLabel]; ok {
-				serviceID = service.ID
+	// only create as many as we need
+	threads := c.MaxThreads
+	if len(serviceQueue) < threads {
+		threads = len(serviceQueue)
+	}
 
-				continue
-			}
+	log.Printf("starting %d threads", threads)
 
-			if len(imageName) > 0 {
-				hasMatch := false
-				for _, imageMatch := range imageName {
-					if strings.HasPrefix(service.Spec.TaskTemplate.ContainerSpec.Image, imageMatch) {
-						hasMatch = true
+	for i := 0; i < threads; i++ {
+		wg.Add(1)
+		go func(key int) {
+			defer wg.Done()
+			log.Printf("starting thread %d", key)
 
-						break
-					}
+			for {
+				if len(serviceQueue) == 0 {
+					log.Printf("exiting thread %d", key)
```

Comment:
IMHO This should only be printed in debug mode.

Reply:
That is a very fair point, I will make those changes